### PR TITLE
Add  transitionOnUserGestures to true on SnackBar for back swipe

### LINKED ISSUE #100782: [IOS]No Hero animation with SnackBar on drag to pop gesture
## Steps to Reproduce

On iOS, while using the drag to pop gesture, the SnackBar appears on both Scaffold during the transition.

1. Execute `flutter run` on the code sample, on an iOS device (or simulator)
2. Click on the _PageTwo_ button
3. Click on the _Show SnackBar_ button
4. Use the drag to pop gesture while the `SnackBar` is still displayed

**Expected results:**

https://user-images.githubusercontent.com/9378033/160168566-3d36aa34-794d-4694-8f47-2e5ea8f330f8.mp4

**Actual results:** 

https://user-images.githubusercontent.com/9378033/160168461-a3af92b2-377f-4570-85b9-b60eab5dd3a5.mp4

We can fix it by adding `transitionOnUserGestures: true,` to the `Hero` created by the `SnackBar`, but I don't know if we should set `true` by default or not.

If we always want this behavior, we can simply set the `Hero` like that:
```dart
return Hero(
  tag: '<SnackBar Hero tag - ${widget.content}>',
  transitionOnUserGestures: true,
  child: snackBarTransition,
);
```

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: PageOne(),
    );
  }
}

class PageOne extends StatelessWidget {
  const PageOne({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          child: const Text('Page Two'),
          onPressed: () {
            Navigator.push(
              context,
              MaterialPageRoute(
                builder: (context) => const PageTwo(),
              ),
            );
          },
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

class PageTwo extends StatelessWidget {
  const PageTwo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Builder(builder: (context) {
        return Center(
          child: ElevatedButton(
            child: const Text('Show SnackBar'),
            onPressed: () {
              ScaffoldMessenger.of(context).showSnackBar(
                const SnackBar(
                  content: Text('SnackBar!'),
                  behavior: SnackBarBehavior.floating,
                ),
              );
            },
          ),
        );
      }),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel stable, 2.10.2, on macOS 12.2.1 21D62 darwin-arm, locale fr-FR)
    • Flutter version 2.10.2 at /Users/romain.rastel/dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 097d3313d8 (il y a 5 semaines), 2022-02-18 19:33:08 -0600
    • Engine revision a83ed0e5e3
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/romain.rastel/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.64.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (3 available)
    • iPhone (mobile)    • 00008030-000139323E86402E            • ios            • iOS 14.8.1 18H107
    • iPhone 13 (mobile) • 7F4ADD7F-5E90-4286-8263-68F02BBF524B • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 99.0.4844.83

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by letsar**
Fixes https://github.com/flutter/flutter/issues/100782

Before : 

https://user-images.githubusercontent.com/9378033/160168461-a3af92b2-377f-4570-85b9-b60eab5dd3a5.mp4

With this PR : 

https://user-images.githubusercontent.com/9378033/160168566-3d36aa34-794d-4694-8f47-2e5ea8f330f8.mp4

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Thank you so much for fixing this! Can you add some documentation to SnackBar about this?

File: packages/flutter/test/material/snack_bar_test.dart [hunk 1]
```diff
@@ -1795,6 +1795,75 @@ void main() {
...
+    // Trigger the swipe.
+    await gesture.moveBy(const Offset(100.0, 0.0));
+
+    // Back gestures should trigger and draw the hero transition in the very same
+    // frame (since the "from" route has already moved to reveal the "to" route).
+    await tester.pump();
+
+    // We should have only one SnackBar displayed on the screen.
+    expect(find.text(snackBarText), findsOneWidget);
+  }, variant: TargetPlatformVariant.only(TargetPlatform.iOS));
```
[the remark this answers is not in the document]
Comment (letsar) [reply packages/flutter/test/material/snack_bar_test.dart:1865]: This transition gesture is supported as soon as the _CupertinoBackGestureDetector is used which is the case by default on iOS. Maybe I should use a CupertinoPageRoute instead to force this scenario.

[the remark this answers is not in the document]
Comment (letsar) [reply packages/flutter/test/material/snack_bar_test.dart]: I made sure the right transition would be use for all platforms, so no need to specify the variant anymore.

**Comment by letsar** [comment]
I've added some documentation but I was not sure how to explain this behavior shortly. Let me know how to improve the documentation if it's not ok.

**Review by Piinks — APPROVED** [review]
LGTM! Thank you!

**Comment by Piinks** [comment]
Oh! We need a secondary review, I will find another reviewer.

**Review by justinmc — APPROVED** [review]
LGTM :+1:

**Comment by Piinks** [comment]
Yeah it looks like it got stuck, I checked and it passed. Merging manually.

**Pull request merged**